### DISCUSSION TITLE
I can't get push notifications to work on iOS. (Requested entity was not found)

### DISCUSSION THREAD

**Discussion by vikjovanov** [invertase/react-native-firebase#4287] [Q&A]
Hi,

I can't get push notifications to work on iOS. On android, everything is OK which confirms that the problem comes from my configuration.

I am able to retrieve a token with iOS (which is 163 characters long) by using the `firebase.messaging().getToken()` function.
However, when I use this token to send a push notification, nothing happens (in the Firebase console). And when my partner (back-end) sends a push notification from our server, he receives a `Failed: 1, success: 0. Requested entity was not found`.

I added the Push notification capability to my project. I created an "APN" key from my Apple Developer account that I added correctly in the Firebase console. I also added a development and production certificate for the push notifications (to make sure the problem couldn't come from the fact that he couldn't send the pushes with the APN key).

The problem is that I'm having trouble debugging from where exactly this can come from!
Is it my iOS configuration? Is it a problem in my code? 

Knowing that I used to use version 6 of GCM in duo with React-native-push-notifications (for interaction management). Since then, I upgraded to Messaging 7 and I can't get it to work anymore.

I have completely removed the entire installation of React-native-push-notifications and React-native-push-notifications-ios because I didn't need it anymore! Maybe I removed too much code?

<details>
  <summary>AppDelegate.h</summary>
  
```objective-c
#import <React/RCTBridgeDelegate.h>
#import <UIKit/UIKit.h>
#import <Firebase.h>

@interface AppDelegate : UIResponder <UIApplicationDelegate, RCTBridgeDelegate>

@property (nonatomic, strong) UIWindow *window;

@end

```

</details>

<details>
<summary>AppDelegate.m</summary>

```objective-c
#import "AppDelegate.h"
#import <React/RCTLinkingManager.h>

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

#import "FirebaseMessaging.h"
#import <RNBootSplash.h>

#ifdef FB_SONARKIT_ENABLED
#import <FlipperKit/FlipperClient.h>
#import <FlipperKitLayoutPlugin/FlipperKitLayoutPlugin.h>
#import <FlipperKitUserDefaultsPlugin/FKUserDefaultsPlugin.h>
#import <FlipperKitNetworkPlugin/FlipperKitNetworkPlugin.h>
#import <SKIOSNetworkPlugin/SKIOSNetworkAdapter.h>
#import <FlipperKitReactPlugin/FlipperKitReactPlugin.h>

static void InitializeFlipper(UIApplication *application) {
  FlipperClient *client = [FlipperClient sharedClient];
  SKDescriptorMapper *layoutDescriptorMapper = [[SKDescriptorMapper alloc] initWithDefaults];
  [client addPlugin:[[FlipperKitLayoutPlugin alloc] initWithRootNode:application withDescriptorMapper:layoutDescriptorMapper]];
  [client addPlugin:[[FKUserDefaultsPlugin alloc] initWithSuiteName:nil]];
  [client addPlugin:[FlipperKitReactPlugin new]];
  [client addPlugin:[[FlipperKitNetworkPlugin alloc] initWithNetworkAdapter:[SKIOSNetworkAdapter new]]];
  [client start];
}
#endif

@implementation AppDelegate



- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [FIRApp configure];
#ifdef FB_SONARKIT_ENABLED
    InitializeFlipper(application);
#endif

  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  
  // UIView *backgroundView = [[[NSBundle mainBundle] loadNibNamed:@"LaunchScreen" owner:self options:nil] firstObject];
  
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                                   moduleName:@"Openup"
                                            initialProperties:nil];

  rootView.backgroundColor = [UIColor clearColor];
  
  
  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];

  [RNBootSplash initWithStoryboard:@"BootSplash" rootView:rootView];
  
  UIApplication.sharedApplication.statusBarHidden = NO;
  return YES;
}



- (BOOL)application:(UIApplication *)application continueUserActivity:(nonnull NSUserActivity *)userActivity
 restorationHandler:(nonnull void (^)(NSArray<id<UIUserActivityRestoring>> * _Nullable))restorationHandler
{
  return [RCTLinkingManager application:application
                   continueUserActivity:userActivity
                     restorationHandler:restorationHandler];
}

- (BOOL)application:(UIApplication *)application
            openURL:(NSURL *)url
            options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{
  return [RCTLinkingManager application:application openURL:url options:options];
}
  
- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

  
  
@end

```
</details>

<details>
<summary>React-native info</summary>

```bash
System:
    OS: macOS 10.14.6
    CPU: (4) x64 Intel(R) Core(TM) i5-3210M CPU @ 2.50GHz
    Memory: 31.88 MB / 16.00 GB
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 10.16.1 - /usr/local/bin/node
    Yarn: Not Found
    npm: 6.13.4 - ~/.npm-global/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 13.2, DriverKit 19.0, macOS 10.15, tvOS 13.2, watchOS 6.1
    Android SDK: Not Found
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.6010548
    Xcode: 11.3.1/11C505 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.4 - /usr/bin/javac
    Python: 3.7.6 - /Users/viktor/Applications/miniconda3/bin/python
  npmPackages:
    @react-native-community/cli: ^4.0.0 => 4.0.1
    react: 16.13.1 => 16.13.1
    react-native: 0.63.2 => 0.63.2
  npmGlobalPackages:
    *react-native*: Not Found
```
</details>

<details>
<summary>Package.json</summary>

```json
...
"dependencies": {
    "@react-native-community/cli": "^4.0.0",
    "@react-native-firebase/app": "^8.4.2",
    "@react-native-firebase/dynamic-links": "^7.5.4",
    "@react-native-firebase/messaging": "^7.8.6",
    "@sentry/react-native": "^1.3.1",
    "@sentry/wizard": "^1.0.2",
    "@types/react": "^16.9.19",
    "react": "16.13.1",
    "react-devtools": "^3.6.3",
    "react-native": "0.63.2",
    "typescript": "^3.9.7",
    ...
  },
...
```
</details>

I can't see what I might have missed! I guess it must already be due to a general understanding problem on my part about how the push notifications works exactly.

Hope you can help me! thank you! 

Viktor

**Comment by mikehardy** [comment]
Not sure what's happening but:
- don't rely on a partner, install the firebase admin node SDK and craft your JSON yourself with a little shell script to send it to the REST end poing for cloud messaging so testing is under your control and repeatable
- make sure you are using a real device. messaging will *only* work on real devices for iOS
- launch the app from Xcode and follow the logs or use Console.app on your mac to do the same

**Reply by vikjovanov** [reply]
Thanks for your answer,

I test only on real device.

I followed what you said with the Console.app.

Although I'm not exactly sure about everything I see, when I send a push notification from the Firebase console, neither Console.app nor the xCode logs tell me anything! 
I really feel that firebase either doesn't trigger APNS or that my config is wrong (although I didn't change anything in the config when I did the shift from V6 to V7).

**Comment by vikjovanov** [comment]
I learned a lot about FCM and how to contact APNS.

As a result, from what I understood, it correctly retrieves the token from APNS and firebase also receives it because it returns an FCM Token to me with the "getToken()" function.

So, I have the feeling that the token returned to me is not valid. As if FCM would generate a bad token (which I find hard to believe).

So I was wondering:
- Is it possible that FCM generates a token for me that is valid in a production context but not valid in a development context?
- Is the error "Requested entity was not found" an error returned by APNS or by FCM?

**Reply by vikjovanov** [reply]
Because I rechecked my configuration several times, followed again the iOS configuration tutorial to make sure I didn't forget anything and my configuration looks really good! So I have a strong feeling that it could just be because I'm working in development mode with development certificates! Unfortunately, I have not yet tested in production to be sure of my assertions.
